File: src/main/java/graph/edge/HalfDuplexChannel.java
package graph.edge;

import graph.vertex.Vertex;

public class HalfDuplexChannel extends AbstractEdge {
    private HalfDuplexChannel pair;

    private HalfDuplexChannel(HalfDuplexChannel channel) {
        super(channel.toNode, channel.fromNode, channel.weight, channel.errors);
        pair = channel;
    }

    public HalfDuplexChannel(Vertex fromNode, Vertex toNode, int weight, double errors) {
        super(fromNode, toNode, weight, errors);
        pair = new HalfDuplexChannel(this);
        connect();
    }

    @Override
    public void setUsed(boolean used) {
        this.used = used;
        pair.used = used;
    }

    @Override
    public void connect() {
        super.connect();
        toNode.addConnection(pair);
    }

    @Override
    public void remove() {
        super.remove();
        toNode.removeConnection(pair);
    }

    @Override
    protected String getDirection() {
        return "<-->";
    }
}

File: src/main/java/mvc/model/channel/HalfDuplexChannel.kt
package mvc.model.channel

import mvc.model.node.Node

class HalfDuplexChannel : AbstractChannel {
    private var pair: HalfDuplexChannel

    override var isUsed = false
        set(used) {
            field = used
            if (pair.isUsed != used)
                pair.isUsed = used
        }

    override val direction = "<-->"

    private constructor(channel: HalfDuplexChannel) : super(channel.toNode, channel.fromNode, channel.weight, channel.errors) {
        pair = channel
    }

    constructor(fromNode: Node, toNode: Node, weight: Int, errors: Double) : super(fromNode, toNode, weight, errors) {
        pair = HalfDuplexChannel(this)
        connect()
    }

    override fun connect() {
        super.connect()
        toNode.addConnection(pair)
    }

    override fun remove() {
        super.remove()
        toNode.removeConnection(pair)
    }
}
